omroneipApp/src/drvOmroneip.cpp

```diff
@@ -0,0 +1,1882 @@
+      bool validDataType = false;
+      for (int t = 0; t < MAX_OMRON_DATA_TYPES; t++)
+      {
+        if (strcmp(words.front().c_str(), omronDataTypes[t].dataTypeString) == 0)
+        {
+          validDataType = true;
+          keyWords.at("dataType") = words.front();
+        }
+      }
+      if (!validDataType)
+      {
+        std::cout << "Datatype invalid" << std::endl;
+        keyWords.at("stringValid") = "false";
+      }
+      words.pop_front();
+    }
+    else if (i == 2)
+    {
+      // Checking for valid sliceSize
+      char *p;
+      if (words.front() == "none"){
+        keyWords.at("sliceSize") = "1";
+      }
+      else {
+        strtol(words.front().c_str(), &p, 10);
+        if (*p == 0)
+        {
+          if (indexable && words.front() != "1")
+          {
+            keyWords.at("sliceSize") = words.front();
+          }
+          else if (words.front() == "0")
+          {
+            keyWords.at("sliceSize") = "1";
+          }
+          else if (words.front() != "1")
+          {
+            std::cout << "You cannot get a slice whole tag. Try tag_name[startIndex] to specify elements for slice" << std::endl;
+            keyWords.at("stringValid") = "false";
+          }
+        }
+        else
+        {
+          std::cout << "Invalid sliceSize, must be integer." << std::endl;
+          keyWords.at("stringValid") = "false";
+        }
+      }
+      words.pop_front();
+    }
+    else if (i == 3)
+    {
+      // Checking for valid offset
+      size_t structIndex = 0; // will store user supplied index into user supplied struct
+      size_t indexStartPos = 0; // stores the position of the first '[' within the user supplied string
+      size_t offset = 0;
+      // user has chosen not to use an offset
+      if (words.front() == "none")
+      {
+        keyWords.at("optimisationFlag") = "dont optimise";
+        keyWords.at("offset") = "0";
+      }
+      else {
+        // attempt to set offset to integer, if not possible then assume it is a structname
+        try
+        {
+          offset = std::stoi(words.front());
+          keyWords.at("optimisationFlag") = "attempt optimisation";
+        }
+        catch(...)
+        {
+          // attempt to split name and integer
+          for (size_t n = 0; n<words.front().size(); n++)
+          {
+            if (words.front().c_str()[n] == '[')
+            {
+              std::string offsetSubstring = words.front().substr(n+1,words.front().size()-(n+1));
+              indexStartPos = n;
+              for (size_t m = 0; m<offsetSubstring.size(); m++)
+              {
+                if (offsetSubstring.c_str()[m] == ']')
+                {
+                  try
+                  {
+                    //struct integer found
+                    structIndex = std::stoi(offsetSubstring.substr(0,m));
+                    keyWords.at("optimisationFlag") = "attempt optimisation";
+                    goto findOffsetFromStruct;
```

Comment:
oh a `goto` - that's a blast from the past. I think these should *not* be used as part of the expected code path - the risk of subtle bugs creeping into future modifications is too great. Please could you think of refactoring this snippet to avoid the goto?

Within the `for` loop you could for example use `break` instead.

Reply:
I distinctly remember feeling very smart when I used that goto and that there was some reason that I couldnt use a break. I think i probably just didn't want to create another flag like indexFound to put around the debugging output and thought this was a neater, although harder to follow, approach.